### PR TITLE
Add plugin sign capabaility

### PR DESCRIPTION
**Pull request by qmuntal**
This PR adds support for using external keys to sign images with the plugins.

### CODE REVIEW

File: internal/cmd/signer.go [hunk 1]
```diff
@@ -14,33 +15,30 @@ import (
 // GetSigner returns a signer according to the CLI context.
 func GetSigner(ctx *cli.Context) (notation.Signer, error) {
 	// read paths of the signing key and its corresponding cert.
```
Comment (gokarnm) [inline internal/cmd/signer.go:17]:
```suggestion
	// Construct a signer from key and cert file if provided as CLI arguments
```

File: internal/cmd/signer.go [hunk 2]
```diff
@@ -14,33 +15,30 @@ import (
...
-		keyPath = key.KeyPath
-		certPath = key.CertificatePath
+	if keyPath := ctx.String(FlagKeyFile.Name); keyPath != "" {
+		certPath := ctx.String(FlagCertFile.Name)
+		return signature.NewSignerFromFiles(keyPath, certPath)
 	}
-
-	// construct signer
-	signer, err := signature.NewSignerFromFiles(keyPath, certPath)
+	key, err := config.ResolveKey(ctx.String(FlagKey.Name))
```
Comment (gokarnm) [inline internal/cmd/signer.go:22]:
```suggestion
    // Construct a signer from preconfigured key pair in config.json
    // if key name is provided as the CLI argument
	key, err := config.ResolveKey(ctx.String(FlagKey.Name))
```

File: internal/cmd/signer.go [hunk 3]
```diff
@@ -14,33 +15,30 @@ import (
...
 	if err != nil {
 		return nil, err
 	}
-	if endpoint := ctx.String(FlagTimestamp.Name); endpoint != "" {
-		if signer.TSA, err = timestamp.NewHTTPTimestamper(nil, endpoint); err != nil {
+	if key.X509KeyPair != nil {
+		return signature.NewSignerFromFiles(key.X509KeyPair.KeyPath, key.X509KeyPair.CertificatePath)
+	}
+	if key.ExternalKey != nil {
```
Comment (gokarnm) [inline internal/cmd/signer.go:29]:
```suggestion
    // Construct a plugin signer if key name provided as the CLI argument
    // corresponds to an external key
	if key.ExternalKey != nil {
```

File: internal/cmd/signer.go [hunk 4]
```diff
@@ -14,33 +15,30 @@ import (
...
 			return nil, err
 		}
+		return &jws.PluginSigner{
+			Runner:       runner,
+			KeyID:        key.ExternalKey.ID,
+			PluginConfig: key.PluginConfig,
+		}, nil
 	}
-	return signer, nil
+	return nil, errors.New("unsupported key")
```
Comment (gokarnm) [inline internal/cmd/signer.go:41]:
```suggestion
	return nil, errors.New("unsupported key, either provide a local key and certificate file paths, or a key name in config.json, check [DOC_PLACEHOLDER] for details")
```

File: pkg/signature/jws.go [hunk 5]
```diff
@@ -24,6 +24,9 @@ func NewSignerFromFiles(keyPath, certPath string) (*jws.Signer, error) {
 	if err != nil {
 		return nil, err
 	}
+	if len(cert.Certificate) == 0 {
+		return nil, errors.New("missing signer certificate chain")
```
Comment (gokarnm) [inline pkg/signature/jws.go:28]:
```suggestion
		return nil, fmt.Errorf("%q does not contain a signer certificate chain", certPath)
```

Comment (gokarnm) [reply pkg/signature/jws.go:28]: Ideally we should push line 22 onward in this method into notation-go, any cert parse/validate should be done in lower layers.

**Review by gokarnm — CHANGES_REQUESTED** [review]

File: pkg/signature/jws.go (hunk 5 above)
Comment (qmuntal) [reply pkg/signature/jws.go]: Agree, but I would leave this change until https://github.com/notaryproject/notation-go/issues/36 is implemented, else we will be moving functions up and down without a clear target.

**Review by gokarnm — APPROVED** [review]

**Review by gokarnm — APPROVED** [review]
LGTM!

**Review by SteveLasker — APPROVED** [review]
LGTM

**Pull request merged**